File: src/week1/leeson2/two-sum.ts
export function twoSum(nums: number[], target: number): number[] {
  const numsWithIndex: { index: number; value: number }[] = Array.from({
    length: nums.length,
  })

  for (let i = 0; i < nums.length; ++i) {
    numsWithIndex[i] = {
      index: i,
      value: nums[i],
    }
  }

  numsWithIndex.sort((a, b) => a.value - b.value)

  let j = nums.length - 1
  for (let i = 0; i < nums.length; ++i) {
    while (i < j && numsWithIndex[i].value + numsWithIndex[j].value > target)
      --j

    if (i < j && numsWithIndex[i].value + numsWithIndex[j].value === target)
      return [numsWithIndex[i].index, numsWithIndex[j].index].sort()
  }

  return []
}

File: src/week2/lesson1/two-sum.cpp
class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> have;

        for (int i = 0; i < nums.size(); ++i) {
            auto it = have.find(target - nums[i]);
            if (it != have.cend()) {
                return vector<int>({it->second, i});
            }

            have[nums[i]] = i;
        }

        return vector<int>({-1, -1});
    }
};